### PR TITLE
[caffe2][nomnigraph] Fix NNPACK relu fusion for inplace Relu

### PR DESCRIPTION
**Pull request by bwasti**

### CODE REVIEW

File: caffe2/python/transformations_test.py [hunk 1]
```diff
@@ -103,3 +103,26 @@ def test_fuseNNPACKConvReluInplaceRelu(self):
...
+        )
+        net.Relu(["Y2"], ["Y2"])
+        addNNPACK(net) # get the NNPACK engine
+        assert (net.Proto().op[0].engine == "NNPACK")
+        fuseNNPACKConvRelu(net)
+        assert (len(net.Proto().op) == 2)
+        has_activation_arg = False
+        for arg in net.Proto().op[0].arg:
+            if arg.name == "activation":
+                assert (arg.s == "Relu")
```
Comment (Maratyszcza) [inline caffe2/python/transformations_test.py]: FYI, Python doesn't require brackets around `assert` condition

File: caffe2/python/transformations_test.py [hunk 2]
```diff
@@ -103,3 +103,26 @@ def test_fuseNNPACKConvReluInplaceRelu(self):
                 has_activation_arg = True
         assert has_activation_arg
         assert net.Proto().op[0].output[0] != net.Proto().op[0].input[0]
+
+    def test_fuseNNPACKConvReluFollowedByMultipleInputOp(self):
```
Comment (Maratyszcza) [inline caffe2/python/transformations_test.py]: Can we have a test involving both in-place and non-in-place Relu?

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h [hunk 3]
```diff
@@ -80,23 +80,47 @@ class Node : public StorageType<T>, public Notifier<Node<T, U>> {
     inEdges.emplace_back(e);
   }
 
+  /// \brief Swaps an old in-edge for a new one.
+  /// \p oldE The old edge to be replaced.
+  /// \p newE A reference to an edge that will be added as an in-edge.
+  void replaceInEdge(EdgeRef oldE, EdgeRef newE) {
+    std::replace(inEdges.begin(), inEdges.end(), oldE, newE);
+    auto oldProducer = oldE->tail();
+    oldProducer->removeOutEdge(oldE);
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h]: I see that the order of edges may change as a result, could it be a problem?

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h [hunk 4]
```diff
@@ -248,17 +272,15 @@ class Graph {
     const auto outEdges = old->getOutEdges();
 
     for (const auto& inEdge : inEdges) {
-      createEdge(inEdge->tail(), newTail);
-    }
-    for (const auto& inEdge : inEdges) {
-      deleteEdge(inEdge);
+      auto producer = inEdge->tail();
+      producer->replaceOutEdge(inEdge, createFloatingEdge(producer, newTail));
+      deleteEdge(inEdge, false);
```
Comment (Maratyszcza) [inline caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h]: Why do you remove the old edge and create a new edge to replace it? Wouldn't updating the tail of the old edge be enough?

**Review by Maratyszcza — APPROVED** [review]
Looks good, please look a the comments before shipping

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h (hunk 3 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h]: this actually fixes the order of edges for the node being replaced but necessarily removes the old outEdge
```
a --> b
```
becomes
```
a -->
c --> b
```
and the dangling `-->` needs to be deleted.  if `a == c` in this case that is still fine

**Pull request merged**